ceph upgrade to v13.3.2: port WindRiver patches

changchengx opened the pull request:
Hi @dtroyer & @oponcea ,
  We've ported all the WR patches to stx-ceph/v13.2.2.
  1. Split the original big patches into small ones.
  2. Remove the WR patch list info in the commit message.
  3. Add some commit message in the generated small patches which are splited from big ones.
  4. ceph cluster works in one StarlingX by doing dedicated-storage deployment with below guide document in virtual environment:
       https://docs.starlingx.io/installation_guide/dedicated_storage.html#dedicated-storage
       Note: doesn't add cinder/glance/swift/nova service. 2 OSDs run-up in cluster.
 5. To check the original patches which these patches are ported from, please check below document
  https://etherpad.openstack.org/p/stx-ceph_upgrade

**Storage ID**

- https://storyboard.openstack.org/#!/story/2003605

**Task ID**

- 28857

Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>
Signed-off-by: Dehao Shang <dehao.shang@intel.com>

LiucIntel committed:
Port: set log date format

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: guard against blkid incorrect output

Upstream: Candidate

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: add journal flush to disk on ceph osd service stop

Upstream: Candidate

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: fix ceph mon started twice

Upstream: Candidate

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Ovidiu Poncea <ovidiu.poncea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: RevertMe: Use user root to run ceph services

Avoid debugging file permission issues when upgrading to Jewel.
This is done to provide the same setup as Hammer in StarlingX.
This commit should be reverted when we decide to enable the ceph user.

Signed-off-by: Daniel Badea <daniel.badea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: spec add monitoring and recovery of hung processes

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: Manage existing ceph journal

Mount existing ceph journal. Add support in ceph-disk for
displaying 'osd uuid' which was removed in Jewel.

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Daniel Badea <daniel.badea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: centos use sysvinit as the init

StarlingX need centos use sysvinit as the init process.

Signed-off-by: Daniel Badea <daniel.badea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: fix ceph-osd: hit core dump when journal flush

Upstream: Candidate (Optional)

This problem may have been fixed by below commit:
    commit a5cd03c643d6cb9074dfd2952cde83435de1b9dd
    Author: wumingqiao <wumingqiao@inspur.com>
However, I still port this patch to fix it with smart pointer.

LevelDBStore creates a CephLevelDBLogger instance it then shares
with libleveldb to write log messages. LevelDBStore instance owns
the pointer to CephLevelDBLogger. However, at exit, when a ceph
flush journal operation is in progress and libleveldb is running
a bacground operation, LevelDBStore destructor deletes the logger
instance before libleveldb instance is deleted (it's a boost::
scoped_ptr). After finishing the background task libleveldb
tries to write a log messages and crashes because the logger was
deleted.

This can be fixed by promoting ceph_logger pointer to boost::
scoped_ptr and rely on automatic destruction of LevelDBSTore
class members to free memory allocated by CephLevelDBLogger
after libleveldb instance is destroyed.

LevelDBStore constructor was moved from .h file to .cc file
to avoid incomplete type error (CephLevelDBLogger).

Signed-off-by: Daniel Badea <daniel.badea@windriver.com>
Signed-off-by: Ovidiu Poncea <ovidiu.poncea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: log rotation is not working for ceph logs

Signed-off-by: Angie Wang <angie.Wang@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: solve osds unexpectedly 'down' after install

'sysvinit' file is required to start a ceph daemon.
For osd daemon on a storage host, this file is created
during 'ceph-disk activate-all', executed from '/etc/init.d/ceph'.
It is possible to have transitory disk I/O errors causing activate to fail
and not create 'sysvinit' file. Give pmon a chance to restart the osd daemon.
For other ceph daemons, 'sysvinit' file creation is triggered differently
(e.g. puppet creates it for monitors).
If daemon type is 'osd', skip checking for 'sysvinit' file presence.

Signed-off-by: Florin Dumitrascu<florin.dumitrascu@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: Add verbose rados & rbd exceptions message

Add type of exception detail to rados & rbd exceptions,
so when failures occur they perculate up to calling
driver (Cinder, Nova, etc.)

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Daniel Badea <daniel.badea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

iotcg commented:
@dtroyer 
If you need to keep the WindRiver patch list info in the commit message, please tell us the format requirement.

LiucIntel committed:
tweak ceph version & gcc check & exclude influx

Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
mgr/restful: check get_metadata before accessing its member

Use variable metadata to check get_metadata return result and
mark the unresponsive server as '[UNKNOWN]' if metadata is None.
When the server isn't responsive, get_metadata will return None
and raise exception when accssing ['hostname'], which will also
result in failing to show other monitors info.

check command:
   curl --silent --insecure --user admin https://localhost:8003/mon
error log:
500 Internal Server Error
  The server encountered an internal error and was unable to complete your request.
  Either the server is overloaded or there is an error in the application.

Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>
Signed-off-by: Kefu Chai <kchai@redhat.com>

LiucIntel committed:
move/rename init-ceph under /etc/rc.d/init.d/

StarlingX use init-ceph(rename: ceph) under
/etc/rc.d/init.d/ to manage ceph

Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

changchengx commented:
@dtroyer @oponcea 
Could you help review this pull request. I have verified it works.

yhu6 commented on stx/mgr-restful-plugin:
```diff
@@ -0,0 +1,321 @@
+    fi
+
+    #ceph config-key set mgr/restful/${user_name}/server_port $PORT
+    ceph config-key set mgr/restful/server_port $PORT
```
"set" might not work with ceph v13.2.2,  need to use "put"

changchengx replied:
“put” is deprecated from luminous(ver12) and ceph prefer to using “set” to do some configuration.
     What’s the problem hit in your side when using “put”?

     I’ve checked the source code: vim cgcs-root/stx/git/ceph/src/mon/ConfigKeyService.cc +215
     “put” and “set” are in the same implementation. The patch is: https://github.com/ceph/ceph/commit/4eb1a518e3ed52103218616cd

changchengx replied:
@yhu6 
I'll use below command to remove the warning:
ceph config set mgr mgr/restful/server_port $PORT

yhu6 replied:
I am good with it!

yhu6 commented on stx/mgr-restful-plugin:
```diff
@@ -0,0 +1,321 @@
+
+    #ceph config-key set mgr/restful/${user_name}/server_port $PORT
+    ceph config-key set mgr/restful/server_port $PORT
+    ceph restful create-self-signed-cert
```
in ~/restful/module.py, "create-self-signed-cert" actually includes a "restart", so you don't need to run the following "ceph restful restart" again.
Instead, to save one more "restart", you could execute "ceph restful create-key admin" prior to "ceph restful create-self-signed-cert", and then make a comment to tell why another "ceph restful restart" is not needed explicitly.

changchengx replied:
@yhu6 Thanks Yong.
restart will be auto triggered in the end of "create-self-signed-cert".
I'll refine it.

yhu6 commented on stx/mgr-restful-plugin:
```diff
@@ -0,0 +1,321 @@
+    ceph restful create-key admin
+#    ceph mgr module enable restful
+    ceph restful restart
+    restful_check
```
After "restful" is restarted either by "ceph restful create-self-signed-cert" or "ceph restful restart", we cannot assure "restful_check" will read out results within expected period of time.
So, you might have to reconsider the implementation of  "restful_check"!!

yhu6 commented on stx/ceph-rest-api:
```diff
@@ -0,0 +1,92 @@
+DESC="ceph-rest-api"
+DAEMON="/usr/bin/ceph-rest-api"
+RUNDIR="/var/run/ceph"
+PIDFILE="${RUNDIR}/ceph-rest-api.pid"
```
this ceps-rest-api.pid path is NOT consistent with the path you specified in "service-mgmt/sm-db-1.0.0/database/create_sm_db.sql"
see https://review.openstack.org/#/c/634120/ in details.
Either way you need to modify one place or another.

changchengx replied:
@yhu6
1) Currently, the system won't execute ceph-rest-api even ceph-rest-api file are built into system 
2) I'm going to delete stx/{ceph-rest-api,ceph-rest-api.service} file from source code.
3) stx/mgr-restful-plugin has set the PID file as /var/run/ceph/mgr/mgr-restful-plugin.pid

Have you hit error in the running system?

yhu6 replied:
just check this path (/var/run/ceph/mgr/mgr-restful-plugin.pid) to make sure it is the same path in sm db file.

LiucIntel committed:
Port: Ceph recovery logic improvements for blocked ops

Signed-off-by: Robert Church <robert.church@windriver.com>
Signed-off-by: Daniel Badea <daniel.badea@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

LiucIntel committed:
Port: Resolve AIO-SX shutdown hang with CEPH ordering hooks

In ceph-10.2.6, the ceph init script uses systemd-run to launch
ceph-mon and ceph-osd services. This generates transient systemd
service files with basic configuration. On node shutdown, ceph is
getting shutdown while it is still in use by containers, and without
unmapping the RBD devices, causing the libceph kernel module to
hang trying to communicate with the ceph monitor.

This update patches the ceph init script to generate systemd
overrides config files for the ceph-mon and ceph-osd that provide
improved ordering during shutdown, as well as a script to run
as part of the docker.service shutdown (by packaging a systemd
override) to unmap the RBD devices. This ordering ensures kubelet
and docker services are shutdown first, then the RBD devices are
cleaned up, followed by the shutdown of the ceph services and
service management (SM). Once kubelet and docker have shut down,
the ceph-preshutdown.sh script is able to cleanly unmount and
unmap the RBD devices and unload the rbd and libceph
kernel modules.

In ceph-11.0.1, the use of systemd-run was replaced with proper
systemd service configuration files. Once ceph is upgraded for
StarlingX, the ordering and cleanup will need to be revisited.

Story: 2003605
Task: 28857

Signed-off-by: Don Penney <don.penney@windriver.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

tingjiec committed:
Archive download instead of git for Ceph submodules

Remove git submodule sync operations since no need, directly copy
and unzip the tarball which is maintainable and stable for build.

Signed-off-by: Chen, Tingjie <tingjie.chen@intel.com>

changchengx committed:
Merge pull request #1 from silverhandy/dev_stx_ceph_13_2_2

Archive download instead of git for ceph submodules - Remove submodules git sync

Commit:
init: skip activating osds

changchengx committed:
Merge pull request #2 from badeadan/dev_stx_ceph_13_2_2

init: skip activating osds

Commit:
ceph: mon down not detected because ceph health detail output changed

"ceph health detail" monitor down status line is indented.
The "status" script does not expect any whitespace at the
beginning of the line so it doesn't properly identify the
monitor as "down".

changchengx committed:
Merge pull request #3 from badeadan/dev_stx_ceph_13_2_2

ceph: mon down not detected because ceph health detail output changed

#TODO
The execute_ceph_cmd function should be removed from src/init-ceph.in.
1) Its implementation is wrong
2) The right implementation is in src/ceph_common.sh
3) src/ceph_common.sh is sourced into src/init-ceph.in
